pkg/models/sigv4.go

```diff
@@ -61,7 +63,19 @@ func (m *SigV4Middleware) signRequest(req *http.Request) (http.Header, error) {
 		return nil, err
 	}
 
-	return signer.Sign(req, nil, "grafana", m.Config.Region, time.Now())
+	if req.Body != nil {
+		// consume entire request body so that the signer can generate a hash from the contents
+		body, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return nil, err
+		}
+		// re-wrap body with a io.ReadCloser for subsequent closure
+		req.Body = ioutil.NopCloser(bytes.NewReader(body))
```

Comment:
What's the point of this? Doesn't `signer.Sign` already do this?

> // Sign will set the request's Body to be the `body` parameter passed in. If
> // the body is not already an io.ReadCloser, it will be wrapped within one

Reply:
The point of this is before we were always calling `signer.Sign(req, nil <body>, ...)` so we would never pass the request body even if it was not nil (so we would never sign the body). With this change, in order to sign the body, we have to read the entire request body, and pass it in separately as an `io.ReadSeeker` so that in the SDK the following code will execute:

```go
var reader io.ReadCloser
if body != nil {
   var ok bool
   if reader, ok = body.(io.ReadCloser); !ok {
      reader = ioutil.NopCloser(body)
   }
}
r.Body = reader
```

Reply:
That wasn't quite my question though. My question is if wrapping req.Body is necessary, considering the SDK method documentation says it does exactly the same.

Reply:
If you're specifically referring to why wrapping it in a no-op closer, the only reason is that's the pattern that I've seen used (also by another sigV4 library which similarly wraps the AWS SDK) when you want to consume a body, but then wrap it up again

Reply:
Yep that's what I'm referring to. What about the SDK documentation for the method saying it will do exactly that? 

> // Sign will set the request's Body to be the body parameter passed in. If
> // the body is not already an io.ReadCloser, it will be wrapped within one

Unless I misunderstand the above, you can drop line 73.